### PR TITLE
Pileup resampling

### PR DESCRIPTION
**Pull request by gaponenko**
This implements the Pileup generator for art-based stopped muon resampling.

### CODE REVIEW

File: EventGenerator/src/DIOGenerator_tool.cc [hunk 1]
```diff
@@ -29,43 +29,54 @@ namespace mu2e {
...
-    {
-
-    }
+    {}
 
+    std::vector<ParticleGeneratorTool::Kinematic> generate() override;
     void generate(std::unique_ptr<GenParticleCollection>& out, const IO::StoppedParticleF& stop) override;
 
-    void setEngine(art::RandomNumberGenerator::base_engine_t& eng) {
+    void finishInitialization(art::RandomNumberGenerator::base_engine_t& eng, const std::string&) override {
```
Comment (brownd1978) [inline EventGenerator/src/DIOGenerator_tool.cc:38]: Why is the material name unused?  DIO spectrum does depend on material, I guess that's a future development?  If so, please add a comment and a TODO flag

File: EventGenerator/src/Pileup_module.cc [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    produces<mu2e::StageParticleCollection>();
+    if(verbosity_ > 0) {
+      mf::LogInfo log("Pileup");
+      log<<"stoppingTargetMaterial = "<<conf().stoppingTargetMaterial()
+         <<", muon lifetime = "<<muonLifeTime_
+         <<", decay fraction = "<<decayFraction_
+         <<std::endl;
+    }
+
+    if(conf().stoppingTargetMaterial() != "Al") {
```
Comment (brownd1978) [inline EventGenerator/src/Pileup_module.cc:107]: It is the individual tools that need to be configured for different materials.  Shouldn't the throw be in finishInitialization?  As far as what this module does (lifetime, decay fraction) many materials besides Al are supported.

File: Mu2eUtilities/src/stoppedMuonList.cc [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    template<class H>
+    std::vector<art::Ptr<SimParticle> > stoppedMuonListImpl(H simh) {
+
+      std::vector<art::Ptr<SimParticle> > res;
+
+      const auto& sims = *simh;
+
+      for(auto i = sims.begin(); i != sims.end(); ++i) {
+        const auto& inpart = i->second;
+        if((inpart.pdgId() == PDGCode::mu_minus) &&
```
Comment (brownd1978) [inline Mu2eUtilities/src/stoppedMuonList.cc:17]: We will eventually need to process stopped mu+, stopped pi-, etc.  is the idea to have dedicated functions for each?  If so, the name stoppedMuonList is ambiguous.  I think a more extensible design is a generic stoppedParticleList function which takes pdgcode and process code as arguments.  Or make this a functor, and provide the pdgcode and process on construction.

**Review by brownd1978 — COMMENTED** [review]
This generally looks good, my suggestions are related to making future extension easier.

**Comment by gaponenko** [comment]
Hi Dave,

On Thu, 10 Jun 2021, David Brown wrote:

> @brownd1978 commented on this pull request.
> 
> This generally looks good, my suggestions are related to making future extension easier.
> 
> >      void generate(std::unique_ptr<GenParticleCollection>& out, const IO::StoppedParticleF& stop) override;
>  
> -    void setEngine(art::RandomNumberGenerator::base_engine_t& eng) {
> +    void finishInitialization(art::RandomNumberGenerator::base_engine_t& eng, const std::string&) override {
> 
> Why is the material name unused?  DIO spectrum does depend on material, I guess that's a future development?  If so, please add a comment and a TODO flag

See below.

> 
> > +    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
> +    , verbosity_{conf().verbosity()}
> +    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
> +    , randFlat_{eng_}
> +    , randExp_{eng_}
> +  {
> +    produces<mu2e::StageParticleCollection>();
> +    if(verbosity_ > 0) {
> +      mf::LogInfo log("Pileup");
> +      log<<"stoppingTargetMaterial = "<<conf().stoppingTargetMaterial()
> +         <<", muon lifetime = "<<muonLifeTime_
> +         <<", decay fraction = "<<decayFraction_
> +         <<std::endl;
> +    }
> +
> +    if(conf().stoppingTargetMaterial() != "Al") {
> 
> It is the individual tools that need to be configured for different materials.  Shouldn't the throw be in finishInitialization?  As far as what this module does (lifetime, decay fraction) many materials besides Al are supported.

That's what I thought, but it does not throw in my testing.  See
https://github.com/Mu2e/Offline/issues/484

In addition to fixing the numbers provided by the
GlobalConstantsService, there is the issue of spectra.  Of course the
DIO spectrum depends on the material, as well as every other spectrum
generated by the Pileup module.  This PR does not touch the existing
implementation of the spectra, which relies on BinnnedSpectrum and:

- assumes that the use provides consistent configuration for individual
  tools to make the spectra consistent

- hardcodes aluminum for some spectra (e.g. MuCapProtonGenerator)
  without saying this explicitly.

IMO the BinnedSpectrum class grew over the years to became one of the
ugliest parts of Mu2e code.  I want to clean this up, which would
involve a significant restructuring.  This would require some not
insignificant amount of effort without changing any of the existing
functionality, so it is not high on my list.  A less error prone
configuration of the Pileup module tools to support non-Al materials
may be an outcome of the code rework.

So yes, future work, but involving more than just the pileup tools.


> 
> > +
> +namespace mu2e {
> +
> +  //================================================================
> +  namespace {
> +
> +    template<class H>
> +    std::vector<art::Ptr<SimParticle> > stoppedMuonListImpl(H simh) {
> +
> +      std::vector<art::Ptr<SimParticle> > res;
> +
> +      const auto& sims = *simh;
> +
> +      for(auto i = sims.begin(); i != sims.end(); ++i) {
> +        const auto& inpart = i->second;
> +        if((inpart.pdgId() == PDGCode::mu_minus) &&
> 
> We will eventually need to process stopped mu+, stopped pi-, etc.  is the idea to have dedicated functions for each?  If so, the name stoppedMuonList is ambiguous.  I think a more extensible design is a generic stoppedParticleList function which takes pdgcode and process code as arguments.  Or make this a functor, and provide the pdgcode and process on construction.


A good point.  I will update this code.

Andrei

**Comment by gaponenko** [comment]
@FNALbuild run build tests

**Review by brownd1978 — APPROVED** [review]
Thanks for making the change.

**Review by AndrewEdmonds11 — APPROVED** [review]
Very nice. Looks good to me. Thanks, Andrei :)

**Pull request merged**